File: progress-bar/progress.py
#!/usr/bin/env python3

import sys
import time
import random

print(f"\x1b]9;4;3;\x1b\\", end="")
sys.stdout.flush()
time.sleep(2)

for i in range(100):
    percent = i + 1

    print(f"\x1b]9;4;1;{percent}\x1b\\", end="")

    filled = percent // 5
    if percent < 100:
        bar = "=" * filled + ">" + " " * (19 - filled)
    else:
        bar = "=" * 20
    print(f"\r[ {bar} ] {percent:>3} / 100% complete", end="")

    sys.stdout.flush()
    time.sleep(random.random() ** 2 * 0.4)

# Clear the progress indicator and move to next line when done
print("\x1b]9;4;0;0\x1b\\", end="")
print()

File: progress-bar/progress.js
#!/usr/bin/env node

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

(async () => {
  process.stdout.write(`\x1b]9;4;3;\x1b\\`);
  await sleep(2000);

  for (let i = 0; i < 100; i++) {
    const percent = i + 1;

    process.stdout.write(`\x1b]9;4;1;${percent}\x1b\\`);

    const filled = Math.floor(percent / 5);
    const bar = percent < 100
      ? "=".repeat(filled) + ">" + " ".repeat(19 - filled)
      : "=".repeat(20);
    process.stdout.write(`\r[ ${bar} ] ${String(percent).padStart(3)} / 100% complete`);

    await sleep(Math.random() ** 2 * 400);
  }

  // Clear the progress indicator and move to next line when done
  process.stdout.write("\x1b]9;4;0;0\x1b\\");
  process.stdout.write("\n");
})();
